salt/states/smartos.py

```diff
@@ -0,0 +1,805 @@
+__virtualname__ = 'smartos'
+
+
+def __virtual__():
+    '''
+    Provides smartos state provided for SmartOS
+    '''
+    return __virtualname__ if 'vmadm.create' in __salt__ else False
+
+
+def _load_config():
+    '''
+    Loads and parses /usbkey/config
+    '''
+    config = {}
+
+    if os.path.isfile('/usbkey/config'):
+        with open('/usbkey/config', 'r') as config_file:
+            for optval in config_file:
+                if optval[0] == '#':
+                    continue
+                if '=' not in optval:
+                    continue
+                optval = optval.split('=')
+                config[optval[0].lower()] = optval[1].strip()
+    log.debug('read /usbkey/config: {0}'.format(config))
+    return config
+
+
+def _write_config(config):
+    '''
+    writes /usbkey/config
+    '''
+    with open('/usbkey/config.salt', 'w') as config_file:
+        config_file.write("#\n# This file was generated by salt\n#\n")
+        for prop in OrderedDict(sorted(config.items())):
+            config_file.write("{0}={1}\n".format(prop, config[prop]))
+
+    if os.path.isfile('/usbkey/config.salt'):
+        try:
+            salt.utils.files.rename('/usbkey/config.salt', '/usbkey/config')
+        except IOError:
+            return False
+        log.debug('wrote /usbkey/config: {0}'.format(config))
+        return True
+    else:
+        return False
+
+
+def _parse_state_config(config, default_config):
+    '''
+    Parse vm_present state config
+    '''
+    for cfg_item in config if config else []:
+        for k in cfg_item:
+            if k in default_config:
+                if isinstance(cfg_item[k], (bool)):
+                    default_config[k] = cfg_item[k]
+                else:
+                    log.warning('config property {0} must be bool, using default: {1}'.format(k, default_config[k]))
+            else:
+                log.warning('config property {0} not one of {1}'.format(k, default_config.keys()))
+
+    return default_config
+
+
+def _parse_vmconfig(config):
+    '''
+    Parse vm_present vm config
+    '''
+    vmconfig = None
+
+    if isinstance(config, (list)):
+        for prop in config:
+            # property
+            if len(prop) == 1 and isinstance(prop, (salt.utils.odict.OrderedDict)):
+                if not vmconfig:
+                    vmconfig = {}
+                for k in prop:
+                    if isinstance(prop[k], (list)):
+                        vmconfig[k] = _parse_vmconfig(prop[k])
+                    else:
+                        vmconfig[k] = prop[k]
+            # instance
+            if len(prop) > 1 and isinstance(prop, (list)):
+                if not vmconfig:
+                    vmconfig = []
+                instance = {}
+                for dataset in prop:
+                    for k in dataset:
+                        instance[k] = dataset[k]
+                vmconfig.append(instance)
+    else:
+        log.error('failed to parse vmconfig')
```

Comment:
Same comment here as above. Let's give the user more contextual information, please.